subscriptions/subscriptions-impl/src/main/java/com/duckduckgo/subscriptions/impl/messaging/SubscriptionMessagingInterface.kt

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright (c) 2023 DuckDuckGo
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.subscriptions.impl.messaging
+
+import android.webkit.JavascriptInterface
+import android.webkit.WebView
+import androidx.core.net.toUri
+import com.duckduckgo.app.di.AppCoroutineScope
+import com.duckduckgo.app.global.DispatcherProvider
+import com.duckduckgo.di.scopes.AppScope
+import com.duckduckgo.js.messaging.api.JsMessage
+import com.duckduckgo.js.messaging.api.JsMessageCallback
+import com.duckduckgo.js.messaging.api.JsMessageHandler
+import com.duckduckgo.js.messaging.api.JsMessageHelper
+import com.duckduckgo.js.messaging.api.JsMessaging
+import com.duckduckgo.js.messaging.api.JsRequestResponse
+import com.duckduckgo.subscriptions.impl.AuthToken
+import com.duckduckgo.subscriptions.impl.JSONObjectAdapter
+import com.duckduckgo.subscriptions.impl.SubscriptionsManager
+import com.squareup.anvil.annotations.ContributesBinding
+import com.squareup.moshi.Moshi
+import javax.inject.Inject
+import javax.inject.Named
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
+import logcat.logcat
+import org.json.JSONObject
+
+interface SubscriptionCallback {
+    fun backToSettings()
+}
+
+@ContributesBinding(AppScope::class)
+@Named("Subscriptions")
+class SubscriptionMessagingInterface @Inject constructor(
+    subscriptionsManager: SubscriptionsManager,
+    private val jsMessageHelper: JsMessageHelper,
+    private val dispatcherProvider: DispatcherProvider,
+    @AppCoroutineScope private val appCoroutineScope: CoroutineScope,
+) : JsMessaging {
+    private val moshi = Moshi.Builder().add(JSONObjectAdapter()).build()
+
+    private lateinit var webView: WebView
```

Comment:
Since this is an AppScoped class, I assume the webview being set here is ActivityScoped, will we have issues with that since the reference is held and seems to not be cleared?

Reply:
Nicely spotted!